OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs

```diff
@@ -105,22 +105,50 @@ public void UpdateMouseover()
 				return;
 			}
 
-			var underCursor = world.ScreenMap.ActorsAt(worldRenderer.Viewport.ViewToWorldPx(Viewport.LastMousePos))
+			var mousePosPx = worldRenderer.Viewport.ViewToWorldPx(Viewport.LastMousePos);
+
+			var underCursor = world.ScreenMap.ActorsAt(mousePosPx)
 				.Where(a => !world.FogObscures(a) && a.HasTrait<IToolTip>())
 				.WithHighestSelectionPriority();
 
+			var frozen = world.ScreenMap.FrozenActorsAt(world.RenderPlayer, mousePosPx)
+				.Where(a => a.TooltipInfo != null && a.IsValid)
+				.WithHighestSelectionPriority();
+
 			if (underCursor != null)
 			{
-				ActorTooltip = underCursor.TraitsImplementing<IToolTip>().First();
-				ActorTooltipExtra = underCursor.TraitsImplementing<IProvideTooltipInfo>();
-				TooltipType = WorldTooltipType.Actor;
-				return;
+				if (frozen != null)
+				{
+					int xDiffUC = underCursor.Bounds.X + underCursor.Bounds.Width / 2 - mousePosPx.X;
+					int yDiffUC = underCursor.Bounds.Y + underCursor.Bounds.Height / 2 - mousePosPx.Y;
+					int xDiffF = frozen.Bounds.X + frozen.Bounds.Width / 2 - mousePosPx.X;
+					int yDiffF = frozen.Bounds.Y + frozen.Bounds.Height / 2 - mousePosPx.Y;
+
+					if (xDiffUC * xDiffUC + yDiffUC * yDiffUC <= xDiffF * xDiffF + yDiffF * yDiffF)
```

Comment:
No, I think the values are always the same if there is both an actor and frozen actor under cursor. So the turret's frozen actor bounds are the same as the actor refinery below it?